File: bin/link_customizations.py
#!/usr/bin/python
#
# link_customizations.py: hey why not do a version of this in python?
# ---------------------------------------------------------------------
#
import os
from os import chdir, getcwd, listdir
from os.path import isfile, isdir, islink
from subprocess import call

def checkForDirectory( dirToCheck ) :
	if isdir( dirToCheck ) :
		print( 'Found a ' + dirToCheck + ' directory, cool.' )
	else:
		print( 'Expecting a "' + dirToCheck + '" directory, but it is not here.' )
		print( 'The current directory is: "' + htdocsDir + '"' )
		print( 'Change to a different directory, preferably one named "htdocs," and try again.' )
		exit( 1 )

def linkLanguageFiles( extension ) :
	print( 'Linking language files for extension ' + extension + '...' )
	rootedSourceDir = htdocsDir + '/' + customizationsDir + '/' + extension + '/language/en-GB'
	destinationDir = htdocsDir + '/' + mainSiteDir + '/language/en-GB'
	languageFiles = listdir( rootedSourceDir )
	### print( 'linkLanguageFiles: rootedSourceDir = ' + rootedSourceDir )
	### print( 'linkLanguageFiles: destinationDir = ' + destinationDir )
	### print( 'languageFiles:' )
	### print( languageFiles )
	for langFile in languageFiles:
		rootedLanguageFile = rootedSourceDir + '/' + langFile
		if( isfile(rootedLanguageFile) ) :
			print( '\tlinking "' + rootedLanguageFile + "\" to\n\t\t\"" + destinationDir + '"' )
			lnCommand = 'cd ' + destinationDir + '; ln -fs ' + rootedLanguageFile + ' .; cd ' + htdocsDir 
		### print( 'lnCommand: ' + lnCommand )
			call( lnCommand, shell=True )

def linkLibrariesVendorDirs( extension ) :
	print( 'Linking libraries/vendor directories for extension ' + extension + '...' )
	rootedSourceDir = htdocsDir + '/' + customizationsDir + '/' + extension + '/libraries/vendor'
	destinationDir = htdocsDir + '/' + mainSiteDir + '/libraries/vendor'
	librariesVendors = listdir( rootedSourceDir )
	### print( 'linkLibrariesVendorDirs test: rootedSourceDir = ' + rootedSourceDir )
	### print( 'linkLibrariesVendorDirs test: destinationDir = ' + destinationDir )
	### print( 'librariesVendors:' )
	### print( librariesVendors )
	for librariesVendorDir in librariesVendors:
		rootedLibrariesVendorDir = rootedSourceDir + '/' + librariesVendorDir
		if( isdir(rootedLibrariesVendorDir) ) :
			print( '\tlinking "' + rootedLibrariesVendorDir + "\" to\n\t\t\"" + destinationDir + '"' )
			lnCommand = 'cd ' + destinationDir + '; ln -fs ' + rootedLibrariesVendorDir + ' .; cd ' + htdocsDir 
		### print( 'lnCommand: ' + lnCommand )
			call( lnCommand, shell=True )

def linkModuleDirs( extension ) :
	print( 'Linking modules directories for extension ' + extension + '...' )
	rootedSourceDir = htdocsDir + '/' + customizationsDir + '/' + extension + '/modules'
	destinationDir = htdocsDir + '/' + mainSiteDir + '/modules'
	moduleDirs = listdir( rootedSourceDir )
	### print( 'linkModuleDirs test: rootedSourceDir = ' + rootedSourceDir )
	### print( 'linkModuleDirs test: destinationDir = ' + destinationDir )
	### print( 'moduleDirs:' )
	### print( moduleDirs )
	for modDir in moduleDirs:
		rootedModuleDir = rootedSourceDir + '/' + modDir
		if( isdir(rootedModuleDir) ) :
			print( '\tlinking "' + rootedModuleDir + "\" to\n\t\t\"" + destinationDir + '"' )
			lnCommand = 'cd ' + destinationDir + '; ln -fs ' + rootedModuleDir + ' .; cd ' + htdocsDir 
		### print( 'lnCommand: ' + lnCommand )
			call( lnCommand, shell=True )

def linkTemplateDirs( extension ) :
	print( 'Linking templates directories for extension ' + extension + '...' )
	rootedSourceDir = htdocsDir + '/' + customizationsDir + '/' + extension + '/templates'
	destinationDir = htdocsDir + '/' + mainSiteDir + '/templates'
	templateDirs = listdir( rootedSourceDir )
	### print( 'linkTemplateDirs test: rootedSourceDir = ' + rootedSourceDir )
	### print( 'linkTemplateDirs test: destinationDir = ' + destinationDir )
	### print( 'templateDirs:' )
	### print( templateDirs )
	for tmplDir in templateDirs:
		rootedTemplateDir = rootedSourceDir + '/' + tmplDir
		if( isdir(rootedTemplateDir) ) :
			print( '\tlinking "' + rootedTemplateDir + "\" to\n\t\t\"" + destinationDir + '"' )
			lnCommand = 'cd ' + destinationDir + '; ln -fs ' + rootedTemplateDir + ' .; cd ' + htdocsDir 
			### print( 'lnCommand: ' + lnCommand )
			call( lnCommand, shell=True )

def linkExtension( extension ) :
	print( 'Linking files in the ' + extension + ' extension...' )
	githubRepoDir = customizationsDir + '/'  + extension
	extensionSubdirs = listdir( githubRepoDir )
	extensionSubdirs.sort()
	for subdirectory in extensionSubdirs:
		if ( subdirectory == 'language' ) :
			linkLanguageFiles( extension )
		elif ( subdirectory == 'libraries' ) :
			linkLibrariesVendorDirs( extension )
		elif ( subdirectory == 'modules' ) :
			linkModuleDirs( extension )
		elif ( subdirectory == 'templates' ) :
			linkTemplateDirs( extension )

customizationsDir = 'customizations'
mainSiteDir = 'joomoowebsites.com'
htdocsDir = getcwd()
exitVal = 0

checkForDirectory( customizationsDir )   # exits if directory not present
checkForDirectory( mainSiteDir )         # exits if directory not present

customizations = listdir( customizationsDir )
customizations.sort()
print( 'Linking the following customizations to the appropriate directories in ' + mainSiteDir + ':' )
print( customizations )

for extension in customizations:
	linkExtension( extension )

exit( exitVal )

File: bin/link_customizations.pl
#!/usr/bin/perl
#
# link_customizations.pl: link customizations in $customizationsDir in cmsDuJour dir.
# -----------------------------------------------------------------------------------
#
use Cwd;
use Cwd 'chdir';

sub linkLanguageFiles( )
{
	my $extension = shift(@_);
	my $rootedLanguageFile;
	my $lnCommand;
	my $rootedSourceDir = $htdocsDir . "/" . $customizationsDir . "/" . $extension . "/language/en-GB";
	my $destinationDir = $htdocsDir . "/" . $cmsDuJour . "/language/en-GB";

	print 'Linking language files for extension ' . extension . "...\n";
	opendir LANGUAGES_DIR, $rootedSourceDir or die "Unable to open \"$rootedSourceDir\" : $!";
	@languageFiles = readdir LANGUAGES_DIR;
	closedir LANGUAGES_DIR;

	### print "rootedSourceDir: \"$rootedSourceDir\"\n";
	### print "destinationDir: \"$destinationDir\"\n";
	### print "Found the following languageFiles: @languageFiles\n";

	foreach $langFile ( @languageFiles )
	{
		$rootedLanguageFile = $rootedSourceDir . "/" . $langFile;
		if ( -f $rootedLanguageFile )
		{
			print "  linking \"$rootedLanguageFile\" to\n\t\"$destinationDir\"\n";
			$lnCommand = "cd $destinationDir; ln -fs $rootedLanguageFile . ; cd $htdocsDir";
		###	print "lnCommand: \"$lnCommand\"\n";
			system $lnCommand;
		}
	}
}

sub linkTemplateDirs( )
{
	my $extension = shift(@_);
	my $tmplDir;
	my $rootedTemplateDir;
	my $lnCommand;
	my $rootedSourceDir = $htdocsDir . "/customizations/" . $extension . "/templates";
	my $destinationDir = $htdocsDir . "/" . $cmsDuJour . "/templates";

	print 'Linking templates directories for extension ' . extension . "...\n";
	opendir TEMPLATES_DIR, $rootedSourceDir or die "Unable to open \"$rootedSourceDir\" : $!";
	@templateDirs = readdir TEMPLATES_DIR;
	closedir TEMPLATES_DIR;

	### print "rootedSourceDir: \"$rootedSourceDir\"\n";
	### print "destinationDir: \"$destinationDir\"\n";
	### print "Found the following templateDirs: @templateDirs\n";

	foreach $tmplDir ( @templateDirs )
	{
		if ( $tmplDir ne "." && $tmplDir ne ".." )
		{
			$rootedTemplateDir = $rootedSourceDir . "/" . $tmplDir;
			if ( -d $rootedTemplateDir )
			{
				print "  linking \"$rootedTemplateDir\" to\n\t\"$destinationDir\"\n";
				$lnCommand = "cd $destinationDir; ln -fs $rootedTemplateDir .; cd $htdocsDir";
			### print "lnCommand: \"$lnCommand\"\n";
				system $lnCommand;
			}
		}
	}
}

sub linkExtension( )
{
	my $extension = shift(@_);
	print "Linking files in the \"$extension\" extension...\n";

	my $githubRepoDir = $customizationsDir . '/' . $extension;
	opendir GITHUB_REPO_DIR, $githubRepoDir or die "Unable to open \"$githubRepoDir\" : $!";
	@extensionSubdirs = readdir GITHUB_REPO_DIR;
	closedir GITHUB_REPO_DIR;
	foreach $subdirectory( sort @extensionSubdirs )
	{
		if ( $subdirectory eq 'language' )
		{
			&linkLanguageFiles( $extension );
		}
		elsif ( $subdirectory eq 'templates' )
		{
			&linkTemplateDirs( $extension );
		}
		else
		{
			print "=======================================================================\n";
			print "Sorry we are not yet equipped to process directories named \"$subdirectory\" .\n";
			print "=======================================================================\n";
		}
	}
}

$customizationsDir = 'customizations';
$cmsDuJour = 'joomla';
$htdocsDir = cwd();
$exitVal = 0;

if ( -d $customizationsDir && -l $cmsDuJour )
{
	print "Found a \"$customizationsDir\" dir and a \"$cmsDuJour\" link, cool.\n";
}
else
{
	print "Expecting a \"$customizationsDir\" dir and a \"$cmsDuJour\" link, but they are not here.\n";
	print "The current directory is: \"$htdocsDir\"\n";
	print "Change to a different directory and try again.\n";
	exit 1;
}

opendir DIR_CUSTOMIZATIONS, "$customizationsDir" or die "Unable to open \"$customizationsDir\" dir. (it was just here?): $!";
@customizations = readdir DIR_CUSTOMIZATIONS;
closedir DIR_CUSTOMIZATIONS;

### print "htdocsDir: \"$htdocsDir\"\n";
### print "@customizations\n";
### $pwd = $ENV{'PWD'};
### print "pwd: \"$pwd\"\n";

foreach $extension ( @customizations )
{
	if ( $extension eq "." || $extension eq ".." )
	{
		next;
	}
	&linkExtension( $extension );
}

exit $exitVal;
